keps/sig-scheduling/20200630-cost-based-scaling-down-of-pods.md

```diff
@@ -0,0 +1,382 @@
+Yet, with some modification another instance of the descheduler (with different set of strategies)
+might be ran in watch mode and rank each pod as it comes.
+Also, once the scheduling framework gets migrated into its own repository,
+scheduling plugins can be vendored as well to provide some of the core scheduling logic.
+
+The pod ranking is best-effort so in case a controller is to delete more than one pod
+it selects all the pods with the highest cost and remove those.
+In case a pod fails to be deleted during the scale down operation and results in resuming the operation in the next cycle,
+it may happen pods get ranked differently and a different set of victim pods gets selected.
+
+Once a pod is removed, ranks of others pods might be required to get re-computed.
+Unless strategies that do not require re-computation are deployed.
+By default, all pods owned by a controller template has to be ranked.
+Otherwise, a controller falls back to each original selection victim logic.
+Resp. it can be configured to wait or back-off.
+Also, the ranking strategies can be configured to target only selected sets of pods.
+Thus, allowing a controller to employ cost based selection only when more sophisticated
+logic is required and available.
+
+During alpha phase, each controller utilizing the pod ranking will feature gate the new logic.
+Starting by utilizing a pod annotation (e.g. `scheduling.alpha.kubernetes.io/cost`)
+which can be eventually promoted to either a field in pod's spec or moved under CRD (see further).
+
+If strategies requiring rank re-computation are employed, it's more practical to define
+a CRD for a pod group and have all the costs in a single place to avoid desynchronization
+of ranks among pods.
+
+#### Phases
+
+Phase 1:
+- add support for strategies which do not need rank re-computation of a pod group
+- only a single strategy can be ran to rank pods (unless threshold based normalization is applied)
+- use annotations to hold a single pod cost
+
+Phase 2A:
+- promote pod cost annotation to a pod status field
+- no synchronization of pods in a pod group, harder support of strategies which require rank re-computation
+
+Phase 2B:
+- use a CRD to hold costs of all pods in a pod group (to synchronize re-computation of ranks)
+- add support for strategies which require rank re-computation
+
+### Option A (field in a pod status)
+
+Store a post cost/rank under pod's status so it can be updated only by component
+who has permission to update pod status.
+
+```go
+// PodStatus represents information about the status of a pod. Status may trail the actual
+// state of a system, especially if the node that hosts the pod cannot contact the control
+// plane.
+type PodStatus struct {
+...
+	// More info: https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle#pod-cost
+	// +optional
+	Cost int `json:"cost,omitempty" protobuf:"bytes,...,opt,name=cost"`
```

Comment:
Is it feasible to have the scheduler record its actual opinion of placement, or to leave room for adding that?

For example: a three-zone cluster, even pod spreading, scaling from 2 to 3. The scheduler places the Pod in the zone with no replica and scores this as “I could not have done better”.

Now schedule 3 more Pods. The first two go leaving this.
- node `a` (zone 1): 1 DaemonSet Pod, 2 app Pods
- node `b` (zone 2): 1 DaemonSet Pod, 1 app Pod, 1 large cluster critical Pod
- node `c` (zone 3): 1 DaemonSet Pod, 2 app Pods

The scheduler considers the 3rd Pod and sees it doesn't fit on Node `b`. The scheduler picks node `c`. The scheduler also records that the scheduling involved a weighted trade-off that is scored at (say) 0.72, because the Pods for the app are not evenly spread.

Reply:
I find the word "cost" very confusing for the intend of this proposal. I think "Penalty" would be more appropriate. Then, that makes it more clear that a 0 penalty means less chance of being evicted.